consensus/hotstuff/committees/consensus_committee.go

```diff
@@ -0,0 +1,206 @@
+	identity, err := c.state.AtBlockID(blockID).Identity(nodeID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get identity for node ID %x: %w", nodeID, err)
+	}
+	if !consensusMemberFilter(identity) {
+		return nil, fmt.Errorf("node with ID %x is not a valid consensus committee member at block %x", nodeID, blockID)
+	}
+	return identity, nil
+}
+
+// LeaderForView returns the node ID of the leader for the given view. Returns
+// the following errors:
+//  * epoch containing the requested view has not been set up (protocol.ErrNextEpochNotSetup)
+//  * any other error indicates an unexpected internal error
+func (c *Consensus) LeaderForView(view uint64) (flow.Identifier, error) {
+
+	// try to retrieve the leader from a pre-computed LeaderSelection
+	id, err := c.precomputedLeaderForView(view)
+	if err == nil {
+		return id, nil
+	}
+	if !errors.Is(err, errSelectionNotComputed) {
+		return flow.ZeroID, err
+	}
+	// we only reach the following code, if we got a errSelectionNotComputed
+
+	// STEP 2 - we haven't yet computed leader selection for an epoch containing
+	// the requested view. We compute leader selection for the current and previous
+	// epoch (w.r.t. the finalized head) at initialization then compute leader
+	// selection for the next epoch when we encounter any view for which we don't
+	// know the leader. The series of epochs we have computed leaders for is
+	// strictly consecutive, meaning we know the leader for all views V where:
+	//
+	//   oldestEpoch.firstView <= V <= newestEpoch.finalView
+	//
+	// Thus, the requested view is either before oldestEpoch.firstView or after
+	// newestEpoch.finalView.
+	//
+	// CASE 1: V < oldestEpoch.firstView
+	// If the view is before the first view we've computed the leader for, this
+	// represents an invalid query because we only guarantee the protocol state
+	// will contain epoch information for the current, previous, and next epoch -
+	// such a query must be for a view within an epoch at least TWO epochs before
+	// the current epoch when we started up. This is considered an invalid query.
+	//
+	// CASE 2: V > newestEpoch.finalView
+	// If the view is after the last view we've computed the leader for, we
+	// assume the view is within the next epoch (w.r.t. the finalized head).
+	// This assumption is equivalent to assuming that we build at least one
+	// block in every epoch, which is anyway a requirement for valid epochs.
+	//
+	next := c.state.Final().Epochs().Next()
+	err = c.prepareLeaderSelection(next)
+	if err != nil {
+		return flow.ZeroID, fmt.Errorf("could not compute leader selection for next epoch: %w", err)
+	}
+	nextCounter, err := next.Counter()
+	if err != nil {
+		return flow.ZeroID, fmt.Errorf("could not get next epoch counter: %w", err)
+	}
+
+	// if we get to this point, we are guaranteed to have inserted the leader
+	// selection for the next epoch
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	nextEpochSelection := c.leaders[nextCounter]
```

Comment:
Must explicitly check here that we are within first/final view of next epoch. Add test for this case.